core/src/test/scala/integration/kafka/server/MetadataRequestBetweenDifferentIbpTest.scala

```diff
@@ -0,0 +1,125 @@
+import kafka.network.SocketServer
+import kafka.server.{BaseRequestTest, KafkaConfig}
+import kafka.utils.TestUtils
+import kafka.zk.ZkVersion
+import org.apache.kafka.common.Uuid
+import org.apache.kafka.common.message.MetadataRequestData
+import org.apache.kafka.common.protocol.Errors
+import org.apache.kafka.common.requests.{MetadataRequest, MetadataResponse}
+import org.junit.jupiter.api.Assertions._
+import org.junit.jupiter.api.Test
+
+import scala.collection.{Map, Seq}
+
+class MetadataRequestBetweenDifferentIbpTest extends BaseRequestTest {
+
+  override def brokerCount: Int = 3
+  override def generateConfigs: Seq[KafkaConfig] = {
+    Seq(
+      createConfig(0, KAFKA_2_8_IV0),
+      createConfig(1, KAFKA_2_8_IV1),
+      createConfig(2, KAFKA_2_8_IV1)
+    )
+  }
+
+  @Test
+  def testTopicIdUnsupported(): Unit = {
+
+    val topic = "topic"
+
+    // Ensure controller version = KAFKA_2_8_IV1, and then create a topic
+    ensureControllerIn(Seq(1, 2))
+    createTopic(topic,  Map(0 -> Seq(1, 2, 0), 1 -> Seq(2, 0, 1)))
+
+    // We can get topicId from the controller
+    val resp1 = sendMetadataRequest(new MetadataRequest(requestData(topic, Uuid.ZERO_UUID), 10.toShort), controllerSocketServer)
+    val topicId = resp1.topicMetadata.iterator().next().topicId()
+    assertNotEquals(Uuid.ZERO_UUID, topicId)
+    assertNotNull(topicId)
+
+    // Send request to a broker whose version=KAFKA_2_8_IV0
+    val resp2 = sendMetadataRequest(new MetadataRequest(requestData(topic, topicId), 10.toShort), brokerSocketServer(0))
+    assertEquals(Errors.UNSUPPORTED_VERSION, resp2.topicMetadata.iterator().next().error())
+  }
+
+  @Test
+  def testUnknownTopicId(): Unit = {
+
+    val topic = "topic"
+
+    // Kill controller and restart until broker 2 become controller
+    ensureControllerIn(Seq(1, 2))
+    createTopic(topic, Map(0 -> Seq(1, 2, 0), 1 -> Seq(2, 0, 1)))
+
+    val resp1 = sendMetadataRequest(new MetadataRequest(requestData(topic, Uuid.ZERO_UUID), 10.toShort), controllerSocketServer)
+    val topicId = resp1.topicMetadata.iterator().next().topicId()
+
+    // We could still get topic metadata by topicId
+    val topicMetadata = sendMetadataRequest(new MetadataRequest(requestData(null, topicId), 10.toShort), controllerSocketServer)
+      .topicMetadata.iterator().next()
+    assertEquals(topicId, topicMetadata.topicId())
+    assertEquals(topic, topicMetadata.topic())
+
+    // Make the broker whose version=KAFKA_2_8_IV0 controller
+    ensureControllerIn(Seq(0))
+
+    // Restart the broker whose version=KAFKA_2_8_IV1, and the controller will send metadata request to it
+    killBroker(1)
+    restartDeadBrokers()
+
+    // Send request to a broker whose version=KAFKA_2_8_IV1 and restarted just now
+    val resp2 = sendMetadataRequest(new MetadataRequest(requestData(topic, topicId), 10.toShort), brokerSocketServer(1))
+    assertEquals(Errors.UNKNOWN_TOPIC_ID, resp2.topicMetadata.iterator().next().error())
```

Comment:
This scenario is a bit of a corner case and maybe not a huge deal either way, but does it make more sense to respond with UNKNOWN_TOPIC_ID or UNSUPPORTED_VERSION here? Maybe either is fine since we can expect UNKNOWN_TOPIC_ID or UNSUPPORTED_VERSION during an upgrade like this anyway.

Reply:
Yes, both ERROR is ok here.